collectors/osd.go

```diff
@@ -750,29 +816,51 @@ func (o *OSDCollector) collectOSDScrubState(ch chan<- prometheus.Metric) error {
 
 func (o *OSDCollector) collectPGRecoveryState(ch chan<- prometheus.Metric) error {
 	for _, pg := range o.pgDumpBrief {
-		if o.initialCollect {
+
+		// We need previous PG state in order to update the metric when a PG has
+		// completed recovery or backfill. Or it could be an empty string if
+		// unknown.
+		prevPGState := o.pgStateCache[pg.PGID]
+		prevNumObjectsRecovered := o.pgObjectsRecoveredCache[pg.PGID]
+		prevBackfillTargets := o.pgBackfillTargetsCache[pg.PGID]
+
+		if prevPGState == "" || strings.Contains(prevPGState, "recovering") || strings.Contains(pg.State, "recovering") ||
+			strings.Contains(prevPGState, "backfilling") || strings.Contains(pg.State, "backfilling") {
 			query, err := o.performPGQuery(pg.PGID)
 			if err != nil {
 				continue
 			}
 
+			o.pgStateCache[pg.PGID] = pg.State
 			o.pgObjectsRecoveredCache[pg.PGID] = query.Info.Stats.StatSum.NumObjectsRecovered
-
-		} else if strings.Contains(pg.State, "recovering") {
-			query, err := o.performPGQuery(pg.PGID)
-			if err != nil {
+			o.pgBackfillTargetsCache[pg.PGID] = query.backfillTargets()
+
+			// There is no previous backfill_targets, and we have just cached
+			// it. Wait for the next time so that we can know the increased
+			// number of objects backfilled for this entire PG and compute the
+			// average increased number of objects backfilled for each OSD in
+			// the backfill_targets.
+			if prevBackfillTargets == nil || len(prevBackfillTargets) == 0 {
 				continue
 			}
 
-			o.pgObjectsRecoveredCache[pg.PGID] = query.Info.Stats.StatSum.NumObjectsRecovered
+			// Average out the total number of objects backfilled to each OSD
+			eachOSDIncrease := float64(o.pgObjectsRecoveredCache[pg.PGID]-prevNumObjectsRecovered) / float64(len(prevBackfillTargets))
```

Comment:
It might be helpful to normalize this to an int since objects recovered should always be whole objects. The computation can have floating point data by when passing the objects recovered information in we should do it at granularity of 1 whole object.

Reply:
Make sense. To normalize, it can be rounded to the nearest integer.